text/3064-bitfield.md

```diff
@@ -0,0 +1,384 @@
+    | - | - | - |
+    | x86_64 | 3 | 1 |
+    | aarch64 | 4 | 4 |
+
+-   Location of bit-fields in the memory layout:
+    
+    ```c
+    struct X {
+        char a;
+        char B:3;
+        char c:2;
+        char d;
+    };
+    ```
+    
+    | Arch | Bits of the memory layout of `X` |
+    | - | - |
+    | x86_64 | `aaaaaaaaa __cccBBB dddddddd` |
+    | mips64 | `aaaaaaaaa BBBccc__ dddddddd` |
+
+-   Dependence of the correct Rust type on the layout of previous fields:
+
+    ```c
+    struct X {
+        char a[N];
+        int b:9;
+        char c;
+    };
+    ```
+
+    ```rust
+    struct X {
+        a: [c_char; N],
+        b: ?,
+        c: c_char,
+    }
+    ```
+
+    On `x86_64`, for `N = 1`, the correct Rust type is `[u8; 2]`. For `N = 3`,
+    the correct Rust type is `u16`. In either case,  the struct requires an
+    additional `repr(align(4))` annotation
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This feature allows you to write C-compatible structs by applying the `bitfield`
+attribute to fields:
+
+```rust
+#[repr(C)]
+struct X {
+    #[bitfield(1)] a: u8,
+    #[bitfield(8)] b: u8,
+    #[bitfield(0)] _: u8,
+    #[bitfield(2)] d: u8,
+}
+```
+
+corresponds to the C struct
+
+```c
+struct X {
+    unsigned char a: 1;
+    unsigned char b: 8;
+    unsigned char  : 0;
+    unsigned char d: 2;
+};
+```
+
+As you can see in the example, unnamed bit-fields in C are written with the
+reserved identifier `_` in Rust. When you have a struct with an unnamed
+bit-field, you cannot access that field. On the other hand, you also do not have
+to define a value for that field when constructing the struct:
+
+```rust
+let x = X { a: 1, b: 1, d: 1 };
+let X { a, b, d } = x;
+```
+
+Just like in C, you cannot take the address of a bit-field:
+
+```rust
+fn f(x: &X) -> &u8 {
+    &x.a // !!! does not compile
+}
+```
+
+The value of the `N` in `bitfield(N)` must be a non-negative integer literal.
+This literal must not be larger than the bit-size of the type of the field.
+
+In debug builds, when you write to a bit-field, Rust performs overflow checks.
+If the value to be stored does not fit in the bit-field, the overflow check
+fails and the operation panics:
+
+```rust
+fn f(x: &mut X) {
+    x.a = 6; // !!! overflow check fails
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The attribute `bitfield(N)` is added. `N` must be an integer literal. `N` is
+called the *width* of the bit-field. This attribute can only be applied to fields
+of a `repr(C)` struct. Such a field is called a *bit-field*. A bit-field must
+have one of the types
+
+- `bool`,
+- `u8`, `u16`, `u32`, `u64`, `u128`, `usize`,
+- `i8`, `i16`, `i32`, `i64`, `i128`, `isize`
+
+modulo type aliases.
+
+The width of a bit-field with type `T` must be in the range `[0,
+bit_width_of(T)]`. For the integer types, `bit_width_of` is defined as `8 *
```

Comment:
Rather than introducing the `bit_width_of` concept, I would suggest handling the cases separately:

The width of a bit-field of type `bool` must be 0 or 1. The width of a bit-field of a numeric type `T` must be in the range `[0, 8 * size_of::<T>()]`.